indico/queries/submission.py

```diff
@@ -0,0 +1,194 @@
+
+    Returns:
+        Submission: Found Submission object
+    """
+
+    query = """
+        query GetSubmission($submissionId: Int!){
+            submission(id: $submissionId){
+                id
+                datasetId
+                workflowId
+                status
+                inputFile
+                inputFilename
+                resultFile
+            }
+        }
+    """
+
+    def __init__(
+        self, submission_id: int,
+    ):
+        super().__init__(
+            self.query, variables={"submissionId": submission_id},
+        )
+
+    def process_response(self, response) -> Submission:
+        return Submission(**(super().process_response(response)["submission"]))
+
+
+class GenerateSubmissionResult(GraphQLRequest):
+    query = """
+        mutation CreateSubmissionResults($submissionId: Int!) {
+            submissionResults(submissionId: $submissionId) {
+                jobId
+            }
+        }
+
+    """
+
+    def __init__(
+        self, submission_id: int
+    ):
+        super().__init__(
+            self.query, variables={"submissionId": submission_id},
+        )
+
+    def process_response(self, response) -> Job:
+        response = super().process_response(response)["submissionResults"]
+        return Job(id=response["jobId"])
+
+
+class SubmissionResult(RequestChain):
+    """
+    Generate a result file for a Submission
+
+    Args:
+        submission_id (int): Id of the submission
+        check_status (str, optional): Submission status to check for.
+            Defaults to any status other than `PROCESSING`
+        wait (bool, optional): Wait until the submission is `check_status`
+            and wait for the result file to be generated. Defaults to False
+        timeout (int or float, optional): Maximum number of seconds to wait before
+            timing out. Ignored if not `wait`. Defaults to 30
+
+    Returns:
+        If `wait`:
+            str: URL to result file that can be retrieved with `RetrieveStorageObject`
+
+        If not `wait`:
+            Job: Job that can be watched for results
+
+    Raises:
+        if `wait`:
+            IndicoTimeoutError: Submission was not `check_status`
+               `or result file Job did not complete within `timeout` seconds.
+        If not `wait`:
+            IndicoInputError: The requested Submission is not `check_status`
+    """
+
+    previous: Submission = None
+
+    def __init__(
+        self,
+        submission_id: int,
+        check_status: str = None,
+        wait: bool = False,
+        timeout: Union[int, float] = 30,
+    ):
+        self.submission_id = submission_id
+        self.wait = wait
+        self.timeout = timeout
+        self.status_check = (
+            partial(eq, check_status) if check_status else partial(ne, "PROCESSING")
+        )
+
+    def requests(self) -> Union[Job, str]:
+        yield GetSubmission(self.submission_id)
+        debouncer = Debouncer(max_timeout=self.timeout)
```

Comment:
Won't this just increase until the next check is timeout time from the previous instead of timing out after timout time? This should just check like a set interval of time apart until done. The debouncer is more for preventing quick retries (albeit I'm not sure when it's really useful to backoff instead of just try every second outside like uploads)